bauble_macro_util/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+                            }
+
+                            Ok(())
+                        }
+                        "attribute" => {
+                            if attribute {
+                                Err(meta.error("duplicate `attribute` attribute"))?
+                            }
+
+                            if !meta.input.is_empty() && !meta.input.peek(Token![,]) {
+                                Err(meta.error("expected no arguments for `attribute` attribute"))?
+                            }
+
+                            attribute = true;
+
+                            Ok(())
+                        }
+                        ident => Err(meta.error(format!("unknown attribute `{ident}`"))),
+                    }
+                })
+                .map_err(|err| err.to_compile_error())?;
+            }
+
+            Ok(FieldAttrs {
+                name: match &field.ident {
+                    Some(ident) => quote! { #ident },
+                    // Tuple structs are constructed with `MyType { 0: val0, 1: val1, ... }` syntax
+                    None => {
+                        let index = Index::from(index);
+                        quote! { #index }
+                    }
+                },
+                ty: match (default, as_default, attribute) {
+                    (Some(_), Some(_), _) => Err(Error::new_spanned(
+                        field,
+                        "field cannot be both `default` and `as_default`",
+                    )
+                    .to_compile_error())?,
+                    (_, Some(_), true) => Err(Error::new_spanned(
+                        field,
+                        "field cannot be both `as_default` and `attribute`",
+                    )
+                    .to_compile_error())?,
+                    (None, Some(as_default), false) => FieldTy::AsDefault {
+                        default: as_default,
+                        ty: &field.ty,
+                    },
+                    (default, None, attribute) => {
+                        let index = Index::from(val_count);
+                        val_count += 1;
+                        FieldTy::Val {
+                            default,
+                            attribute,
+                            index,
+                            ty: &field.ty,
+                        }
+                    }
+                },
+            })
+        })
+        .collect::<Result<Vec<_>, _>>()
+    {
+        Ok(fields) => fields,
+        Err(err) => return err,
+    };
+
+    // Generate functions for default values
+    let defaults = fields.iter().filter_map(|field| match &field.ty {
+        FieldTy::Val {
+            default: Some(default),
+            ty,
+            ..
+        }
+        | FieldTy::AsDefault { default, ty } => {
+            let name = Ident::new(&format!("default_{}", field.name), Span::call_site());
```

Comment:
There's a macro to make this less tedious: https://docs.rs/quote/1.0.28/quote/macro.format_ident.html